src/System.Net.Security/src/System/Net/_SslSessionsCache.cs

```diff
@@ -0,0 +1,201 @@
+        //
+        internal static void CacheCredential(SafeFreeCredentials creds, byte[] thumbPrint, int allowedProtocols, EncryptionPolicy encryptionPolicy)
+        {
+            GlobalLog.Assert(creds != null, "CacheCredential|creds == null");
+            if (creds.IsInvalid)
+            {
+                GlobalLog.Print("CacheCredential() Refused to cache an Invalid Handle = " + creds.ToString() + ", Current Cache Count = " + s_CachedCreds.Count);
+                return;
+            }
+
+            object key = new SslCredKey(thumbPrint, allowedProtocols, encryptionPolicy);
+
+            SafeCredentialReference cached = s_CachedCreds[key] as SafeCredentialReference;
+
+            if (cached == null || cached.IsClosed || cached._Target.IsInvalid)
+            {
+                lock (s_CachedCreds)
+                {
+                    cached = s_CachedCreds[key] as SafeCredentialReference;
+
+                    if (cached == null || cached.IsClosed)
+                    {
+                        cached = SafeCredentialReference.CreateReference(creds);
+
+                        if (cached == null)
+                        {
+                            // Means the handle got closed in between, return it back and let caller deal with the issue.
+                            return;
+                        }
+
+                        s_CachedCreds[key] = cached;
+                        GlobalLog.Print("CacheCredential() Caching New Handle = " + creds.ToString() + ", Current Cache Count = " + s_CachedCreds.Count);
+
+                        //
+                        // A simplest way of preventing infinite cache grows.
+                        //
+                        // Security relief (DoS):
+                        //     A number of active creds is never greater than a number of _outstanding_
+                        //     security sessions, i.e. SSL connections.
+                        //     So we will try to shrink cache to the number of active creds once in a while.
+                        //
+                        //    We won't shrink cache in the case when NO new handles are coming to it.
+                        //
+                        if ((s_CachedCreds.Count % c_CheckExpiredModulo) == 0)
+                        {
+                            DictionaryEntry[] toRemoveAttempt = new DictionaryEntry[s_CachedCreds.Count];
+                            s_CachedCreds.CopyTo(toRemoveAttempt, 0);
+
+                            for (int i = 0; i < toRemoveAttempt.Length; ++i)
+                            {
+                                cached = toRemoveAttempt[i].Value as SafeCredentialReference;
+
+                                if (cached != null)
+                                {
+                                    creds = cached._Target;
+                                    cached.Dispose();
```

Comment:
Why not just use AddRef/Release on the underlying handle directly and get rid of the middle-man?  The notion of caching here makes it feel like you can end up wanting to share, and I'm not sure how the one-shot wrapper really accomplishes that.